pyam/iiasa.py

```diff
@@ -17,51 +19,110 @@
+        msg = 'Could not read credentials: {}\n{}'.format(
+            creds, str(e))
+        raise type(e)(msg)
+
+    headers = {'Accept': 'application/json',
+               'Content-Type': 'application/json'}
+    data = {'username': user, 'password': pw}
+    url = '/'.join([_BASE_URL, 'login'])
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+    if not response.ok:  # auth failed, provide user a nice message
+        raise RuntimeError('Login failed for user: {}'.format(user))
+
+    return response.json()
 
 
 class Connection(object):
     """A class to facilitate querying an IIASA scenario explorer database"""
 
-    def __init__(self, name):
+    def __init__(self, name=None, creds=None):
         """
         Parameters
         ----------
-        name : str
+        name : str, optional
             A valid database name. For available options, see
             valid_connection_names().
+        creds : list-like or dict, optional
+            An ordered container with entries of 'username' and 'password',
+            or a dictionary with the same keys.
         """
-        valid = valid_connection_names()
+        self._token = _get_token(creds)
+
+        # connect if provided a name
+        self._connected = None
+        if name:
+            self.connect(name)
+
+    @property
+    @lru_cache()
+    def valid_connections(self):
+        url = '/'.join([_BASE_URL, 'applications'])
+        headers = {'Authorization': 'Bearer {}'.format(self._token)}
+        response = requests.get(url, headers=headers).json()
+        valid = [x['name'] for x in response]
+        return valid
+
+    def connect(self, name):
+        # TODO: deprecate in next release
+        if name == 'iamc15':
+            warnings.warn(
+                'The name `iamc15` is deprecated and will be removed in the ' +
+                'next release. Please use `IXSE_SR15`.'
+            )
+            name = 'IXSE_SR15'
+
+        valid = self.valid_connections
+        if len(valid) == 0:
+            raise RuntimeError(
+                'No valid connections found for the provided credentials.'
+            )
+
         if name not in valid:
-            raise ValueError('{} is not a valid name. Choose one of {}'.format(
-                name, valid))
+            msg = """
+            {} not recognized as a valid connection name.
+            Choose from one of the supported connections for your user: {}.
+            """
+            raise ValueError(msg.format(name, valid))
 
-        logger().info(
-            'You are connected to the {} {}. Please cite as:\n\n{}'
-            .format(name, 'scenario explorer', _CITATIONS[name])
-        )
+        url = '/'.join([_BASE_URL, 'applications', name, 'config'])
+        headers = {'Authorization': 'Bearer {}'.format(self._token)}
+        response = requests.get(url, headers=headers).json()
+        idxs = {x['path']: i for i, x in enumerate(response)}
```

Comment:
Maybe I misunderstand how it works, but the response should contain json similar to following one
```
[
    {
        "name": "IXSE_SETNAV",
        "scheme": "IXMP Scenario Explorer SPA UI",
        "config": [
            {
                "path": "productName",
                "type": "STRING",
                "value": "SET-Nav Scenario Explorer"
            },
            {
                "path": "uiUrl",
                "type": "STRING",
                "value": "https://data.ene.iiasa.ac.at/set-nav"
            }
        ]
    },
    {
        "name": "IXSE_SR15",
        "scheme": "IXMP Scenario Explorer SPA UI",
        "config": [
            {
                "path": "productName",
                "type": "STRING",
                "value": "IAMC 1.5°C Scenario Explorer hosted by IIASA"
            },
            {
                "path": "uiUrl",
                "type": "STRING",
                "value": "https://data.ene.iiasa.ac.at/iamc-1.5c-explorer"
            }
        ]
    },
    {
        "name": "IXSE_SR15_QA",
        "scheme": "IXMP Scenario Explorer SPA UI",
        "config": [
            {
                "path": "productName",
                "type": "STRING",
                "value": "IAMC 1.5°C Scenario Explorer hosted by IIASA (QA)"
            },
            {
                "path": "uiUrl",
                "type": "STRING",
                "value": "https://data.ene.iiasa.ac.at/IXMPtest/scenario-explorer-qa"
            }
        ]
    },
    {
        "name": "IXSE_TEST_PUBLIC",
        "scheme": "IXMP Scenario Explorer SPA UI",
        "config": [
            {
                "path": "productName",
                "type": "STRING",
                "value": "Scenario Explorer (sandbox)"
            },
            {
                "path": "uiUrl",
                "type": "STRING",
                "value": "https://data.ene.iiasa.ac.at/ixmp-explorer-sandbox"
            }
        ]
    }
]
```
How does it select specific app?

Reply:
Sorry, I overlook name in `        url = '/'.join([_BASE_URL, 'applications', name, 'config'])`

Reply:
Yeah, this queries information after `connect()` is called, which aligns it to a specific application